File: templates/javascript/v1/command/commands.js
export class TextBuffer {
  constructor() {
    this.buffer = "";
  }

  append(text) {
    this.buffer += text;
  }

  snapshot() {
    return this.buffer;
  }
}

export class AppendTextCommand {
  constructor(buffer, text) {
    this.buffer = buffer;
    this.text = text;
  }

  execute() {
    this.buffer.append(this.text);
    return this.buffer.snapshot();
  }
}

export class CommandInvoker {
  constructor() {
    this.history = [];
  }

  run(command) {
    this.history.push(command);
    return command.execute();
  }

  historySize() {
    return this.history.length;
  }
}

File: templates/python/v1/command/commands.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Protocol


class Command(Protocol):
    def execute(self) -> str:
        ...


@dataclass(slots=True)
class TextBuffer:
    buffer: str = ""

    def append(self, text: str) -> None:
        self.buffer += text

    def snapshot(self) -> str:
        return self.buffer


@dataclass(slots=True)
class AppendTextCommand:
    buffer: TextBuffer
    text: str

    def execute(self) -> str:
        self.buffer.append(self.text)
        return self.buffer.snapshot()


@dataclass(slots=True)
class CommandInvoker:
    history: List[Command] = field(default_factory=list)

    def run(self, command: Command) -> str:
        self.history.append(command)
        return command.execute()

    def history_size(self) -> int:
        return len(self.history)
